### ISSUE TITLE
[SIG-scale] Work queue performance

### ISSUE THREAD

**Issue by rthallisey** [kubevirt/kubevirt#6055]
### Summary
Create n VMIs, wait until they are all running, delete n VMIs, wait until they are all deleted, sleep 5s, and repeat. Scenarios creating 10, 20, 40, 60, 80, 100, 200, 300 VMIs shows high work queue latency, an exponential increase in work queue depth, and a lot of unfinished work/long running processes.

- Work queue latency is high (8s!) even from 10 VMIs.
- Work queue depth is growing very fast
- High amounts of unfinished work and long running processes

### KubeVirt control plane metrics
<img width="468" alt="work-queue-latency-5982" src="https://user-images.githubusercontent.com/6795213/125938756-f604c8ee-8bf4-4061-ac19-1b580e3c57ac.png">

<img width="468" alt="work-queue-depth-5982" src="https://user-images.githubusercontent.com/6795213/125938765-b3004834-6153-4494-bfd9-73c3a811dca9.png">

<img width="468" alt="work-queue-work-5982" src="https://user-images.githubusercontent.com/6795213/125938775-6b7dc2aa-5670-4dd0-815f-45685b4d0304.png">

### Metrics
| Name | Description | Metric |
|--|---|--|
| work queue add rate | Total number of adds handled by a workqueue | `sum(rate(kubevirt_workqueue_adds_total{job=~".*kubevirt.*"}[5m])) by (instance, name)`|
| work queue latency | How long an item stays in a workqueue | `histogram_quantile(0.99, sum(rate(kubevirt_workqueue_queue_duration_seconds_bucket{job=~".*kubevirt.*", cluster="$cluster"}[5m])) by (instance, name, le))` |
| work queue depth | Current depth of a workqueue | `sum(rate(kubevirt_workqueue_depth{job=~".*kubevirt.*"}[5m])) by (instance, name)`|
| unfinished work | Time in seconds taken for the work in progress and has not been observed by work_duration. Large values indicate stuck threads | `sum(increase(kubevirt_workqueue_unfinished_work_seconds{job=~".*kubevirt.*", cluster="$cluster"}[5m])) by (instance, name)` |
| longest running processor | Time in seconds for which the longest running processor for workqueue is running | `sum(rate(kubevirt_workqueue_longest_running_processor_seconds{job=~".*kubevirt.*", cluster="$cluster"}[5m])) by (instance, name)` |

/kind bug

**Comment by rthallisey** [comment]
Gather work queue perf after https://github.com/kubevirt/kubevirt/pull/6101 merges and see if there's a difference

**Comment by rthallisey** [comment]
@marceloamaral did some new [experiments](https://docs.google.com/document/d/1jGQvYaxh5BKSrg6WKfeYGQ6957Ext3dP5lVhaSMbEHk/edit#).  

![vm-count](https://user-images.githubusercontent.com/6795213/133817870-6854ab0d-be42-487a-87ee-bf7fcc8ebb74.png)

Unfinished work is still high.  This is the longest running key from the work queue.
![unfinished-work](https://user-images.githubusercontent.com/6795213/133817080-4596e7d3-256f-4fd5-a860-ba971e606cfc.png)

Virt-controller node work queue seems really busy compared to other work queues.
![work-queue-latency](https://user-images.githubusercontent.com/6795213/133817878-e6b6890c-fe6a-4c0e-81c2-79542ac00584.png)

**Comment by zcahana** [comment]
I've proposed some updates to the control plane dashboard, including to the work queue metrics panels. If interested, please take a look at kubevirt/monitoring#61.

**Comment by marceloamaral** [comment]
Thanks @zcahana for updating the dashboard in kubevirt/monitoring#61
I've rerun the experiments with the lastest KubeVirt version from the main branch and used the new dashboard.

The vmi creation time from 600 to 800 is still weak scaling
![image](https://user-images.githubusercontent.com/5133121/142839374-7eb93aa8-ce38-4791-bb13-60c87ad76bbb.png)

Unfinished work and longest running key is still high, even after fixing the promQL (i.e., removing `rate()` from Gauge metrics)
![image](https://user-images.githubusercontent.com/5133121/142839253-acf21d09-e955-4862-a78e-ec309272514c.png)

You can find more details in the following links: https://snapshot.raintank.io/dashboard/snapshot/Et9ReahFAD5o9RYwYBmrvB8HDWH5MmMG?orgId=2
https://snapshot.raintank.io/dashboard/snapshot/HHd0PP7i3rRsLemfk1GYyBvsykmWMF7r?orgId=2

**Comment by marceloamaral** [comment]
The workqueue performance is directly related to client-go rate limiter configuration, i.e., QPS and Bust.

In an experiment, I increased the virt-controller rate limiter from 5/10 to 200/400 QPS/Burst which improved the overall workqueue performance.

Consider the following figure, left has a higher rate limiter and right the default rate limiter.

<img width="1326" alt="image" src="https://user-images.githubusercontent.com/5133121/159486575-f3436336-c053-474c-8b72-a3d7f65f9509.png">

You can see that after increasing the virt-controller rate limiter configuration, the VM creation time (ie going into ready condition) has dramatically decreased. For example, when creating 950 VMs, without rate limiter configuration, the average creation time was ≅17min, but after updating the virt-controller rate limiter configuration, the average VM creation time dropped to less than 2min.

The throughput improvement can also be seen in the workqueue.
<img width="1334" alt="image" src="https://user-images.githubusercontent.com/5133121/159487000-654a5fd6-b0d0-44cc-89a0-8e9a90f63186.png">
There are more add operations in the workqueue. Also, the time that a key stays in the Queue and the processing time (work time) have also decreased, as well as the unfinished work and long running processor time. Showing that the threads are processing the keys faster.

In conclusion, increasing the rate limiter is critical for good performance in relation to VM build latency.

**Comment by marceloamaral** [comment]
As a next steps, I will test it with different QPS/Burst configurations
